### DISCUSSION TITLE
Compiling raylib with tcc (Tiny C Compiler), emmintrin.h not found

### DISCUSSION THREAD

**Discussion by GroundhogGrafix** [raysan5/raylib#2994] [General]
PROBLEM:
I am attempting to compile raylib using tcc but I'm encountering the following error:

D:/raylib-4.5.0/src/external/msf_gif.h:249: error: include file 'emmintrin.h' not found

ATTEMPTED SOLUTION:
I copied 'emmintrin.h' from my MinGW64 installation to tcc's include directory. This caused another 'not found' file error for 'xmmintrin.h'. After copying xmmintrin.h in to tcc's include folder, I had the same error with the file 'mmintrin.h'. Now all 3 files have been copied in, I receive this error: 'tcc/include/mmintrin.h:66: error: implicit declaration of function '__builtin_ia32_emms''.

POTENTIAL ISSUES:
Once I get raylib compiled with tcc, what known issues are there? I have read that there may be audio problems.

MY SETUP:
tcc version: 0.9.27
OS: Windows 11 x64

CONCLUSION:
Since MinGW's headers are not right for this solution and if there are audio problems like I have read about online. What about compiling a dll shared library using MinGW gcc and importing the dll in to a tcc def file using the compiler flag -impdef?

**Comment by Peter0x44** [comment]
> D:/raylib-4.5.0/src/external/msf_gif.h:249: error: include file 'emmintrin.h' not found

The correct solution to this is not to copy a header for the intrinsics (that doesn't make them work), but instead to define MSF_GIF_NO_SSE2.
Then, it shouldn't attempt to include the header at all.

Maybe there are further problems - I think TCC hasn't been tested in a while.

**Reply by GroundhogGrafix** [reply]
You are right there are more problems but that define at least helped. Thank you for that.

Now the next problem is:
[  6%] Linking C static library libglfw3.a
[  6%] Built target glfw
[  6%] Building C object raylib/CMakeFiles/raylib.dir/rcore.c.obj
d:/tcc/include/math.h:217: error: unknown constraint 't'
gnumake[2]: *** [raylib\CMakeFiles\raylib.dir\build.make:74: raylib/CMakeFiles/raylib.dir/rcore.c.obj] Error 1

**Reply by Peter0x44** [reply]
That's strange. I would try building the latest tcc yourself, the last release was long ago and has many problems.

I just tried with the latest raylib git, and latest tcc git, and it worked fine for me, building both raylib and an executable linking to it.
No source modifications required.
I am not using windows, however. I can give that a deeper look later.

**Reply by GroundhogGrafix** [reply]
I downloaded tiny c from here: https://github.com/TinyCC/tinycc

It says it's unofficial but it at least has some recent work done to it. Recent being within the last 2 weeks. After I built it, I got a little further. :)

Now the problem is:
[  8%] Building C object raylib/CMakeFiles/raylib.dir/raudio.c.obj
In file included from D:/raylib-4.5.0/src/raudio.c:137:
In file included from d:/tcc/include/winapi/windows.h:17:
d:/tcc/include/_mingw.h:136: warning: NOSERVICE redefined
In file included from D:/raylib-4.5.0/src/raudio.c:137:
In file included from d:/tcc/include/winapi/windows.h:17:
d:/tcc/include/_mingw.h:137: warning: NOMCX redefined
D:/raylib-4.5.0/src/raudio.c:154: error: include file 'objbase.h' not found
gnumake[2]: *** [raylib\CMakeFiles\raylib.dir\build.make:152: raylib/CMakeFiles/raylib.dir/raudio.c.obj] Error 1

**Reply by GroundhogGrafix** [reply]
Well I am glad to hear that you did get it to compile successfully. I am not very familiar with modern programming tools, I mostly do K&R style c for older platforms, DOS and 6502 based systems. I'm trying to use CMake to setup the make file. Maybe I'm doing something wrong there. Perhaps there is another approach to this that I should take?

**Reply by Peter0x44** [reply]
> It says it's unofficial but it at least has some recent work done to it. Recent being within the last 2 weeks. After I built it, I got a little further. :)

The repo is unofficial, not the code present in it. It's just a mirror of the main one.

I tried myself and can reproduce this same problem too. objbase.h isn't present anywhere in the source of tcc, so I'm not really sure how to address that properly.
CMake isn't required, but it is also not causing that issue, as far as I can tell.

Consider opening an issue for this, I guess. Would be nice to get other suggestions and people trying.

**Reply by EmmanuelOga** [reply]
I suspect this should help:
https://stackoverflow.com/questions/24085935/compiling-a-file-with-tcc-on-windows-with-a-library-from-c99

... which links to tcc docs:

```
    Header Files:
    -------------
    The system header files (except _mingw.h) are from the MinGW
    distribution:

	http://www.mingw.org/

    From the windows headers, only a minimal set is included.  If you need
    more,  get MinGW's "w32api" package.  Extract the files from "include"
    into your "tcc/include/winapi" directory.
```

... so you'll probably want to install https://www.mingw-w64.org/ and then copy whatever include files are missing to the place where TCC expect them, or maybe add an include param (-Idir) to the mingw folder that contains objbase.h.

**Reply by EmmanuelOga** [reply]
BTW, found a couple of example Docker files for running mingw:

https://github.com/pdutton/container-golang-ming/blob/WINDOWS/Dockerfile
https://github.com/inkydragon/msys2-mingw-w64/blob/main/Dockerfile

... so you could potentially modify those to run TCC in docker (I suspect pacman install TCC should work!). These need to be run from a Windows host.

The caveat: it generates a 2gb image or so, but you would only need to do that once. The advantage of TCC is a little bit its tiny size but more than that, the fact that it compiles C code so fast! So maybe users wouldn't be bothered by the huge image size... although it _is_ a pity that the image is 1000 times bigger than the size of TCC itself :-(.

There's a nano server windows image that is smaller (~600mb?), but I'm not sure it could run mingw, fwiw. But since the mingw compiler doesn't really need to run, just provide the includes, maybe that nano image could work with TCC after all.

The advantage is that anybody with docker installed could run TCC with a fully working environment, assuming you get to make it work.

Finally, in an unrelated note (since I brought up Docker), https://github.com/dockcross/dockcross could be interesting to be able to compile raylib apps from Windows or Mac to android, linux, windows or wasm. 

P.S.: I don't run windows a lot these days so I haven't tried my own idea, but leaving this here in case somebody wants to try.

**Comment by sleeptightAnsiC** [comment]
I don't believe this is a problem with MinGW or with TCC's distribution specific to Windows,
I'm also noticing this issue on Linux with TCC 0.9.28 (git trunk) and latest Raylib:
```
~/raylib/src (master) $ make CC=tcc                                                                                                                                                          
tcc -c rtextures.c -Wall -D_GNU_SOURCE -DPLATFORM_DESKTOP_GLFW -DGRAPHICS_API_OPENGL_33 -Wno-missing-braces -Werror=pointer-arith -fno-strict-aliasing -std=c99 -fPIC -O1 -Werror=implicit-function-declaration -D_GLFW_X11  -I.  -Iexternal/glfw/include                                                                                                                                                             
In file included from rtextures.c:222:                                                                                                                                                                     
external/stb_image_resize2.h:869: warning: #pragma STDC ignored                                                                                                                                            
In file included from rtextures.c:222:                                                                                                                                                                     
external/stb_image_resize2.h:1279: error: include file 'emmintrin.h' not found                                                                                                                             
make: *** [Makefile:745: rtextures.o] Error 1  
```

I believe this might be an issue with `stb_image_resize2` which naively assumes that compiler comes with `emmintrin.h` without doing any compiler checks, it just checks CPU architecture (it even allows non-SSE2 CPUs):
https://github.com/raysan5/raylib/blob/0275d13aa84f8a464b5246abbef9205b94c73c5e/src/external/stb_image_resize2.h#L392-L396
 This header is in fact not portable and it holds compiler specific code. Not sure if anything can be done on TCC side here, maybe it can emulate this functionality.

I'll try to mess with this a bit and forward this issue to either STB repo or TCC-devel mailing list.

In the meantime, the quick'n'dirty workaround is checking for TCC with `__TINYC__` predef. Here's the patch:
```patch
diff --git a/src/external/stb_image_resize2.h b/src/external/stb_image_resize2.h
index 2f26274..6a3ebc5 100644
--- a/src/external/stb_image_resize2.h
+++ b/src/external/stb_image_resize2.h
@@ -390,6 +390,8 @@ typedef uint64_t stbir_uint64;
 #endif
 #endif
 
+#ifndef __TINYC__
+
 #if defined(_x86_64) || defined( __x86_64__ ) || defined( _M_X64 ) || defined(__x86_64) || defined(_M_AMD64) || defined(__SSE2__) || defined(STBIR_SSE) || defined(STBIR_SSE2)
   #ifndef STBIR_SSE2
     #define STBIR_SSE2
@@ -438,6 +440,8 @@ typedef uint64_t stbir_uint64;
 #endif
 #endif
 
+#endif  // __TINYC__
+
 #ifndef STBIRDEF
 #ifdef STB_IMAGE_RESIZE_STATIC
 #define STBIRDEF static
```
Not sure if it works though. It compiles Raylib just fine but it may cause some linking errors when building the game or some issues with the library itself.

A similar issue from different project - https://github.com/linux-test-project/ltp/issues/980 - they solved it with checking for `HAVE_EMMINTRIN_H` which I assume was predefined by their build system so this approach would not work here.

**Reply by raysan5** [reply]
@sleeptightAnsiC thanks for the further investigation and fordwarding the issue to respective repos. Definitely having raylib working on TCC is really nice.

**Reply by sleeptightAnsiC** [reply]
So I've reported this in STB, here https://github.com/nothings/stb/issues/1738
and right after reporting realized there are `STBIW_NO_SIMD` and `STBI_NO_SIMD` macros...
We can easily fix this on raylib side. I've just added the PR here: https://github.com/raysan5/raylib/pull/4707
This is also how the rest of library is dealing with this.
cc @raysan5 @GroundhogGrafix
